Math/NumberTheory/Moduli/Sqrt.hs

```diff
@@ -13,30 +13,78 @@
 {-# LANGUAGE CPP          #-}
 
 module Math.NumberTheory.Moduli.Sqrt
-  ( sqrtModP
-  , sqrtModPList
-  , sqrtModP'
-  , tonelliShanks
-  , sqrtModPP
-  , sqrtModPPList
+  ( sqrtModList
+  , sqrtModMaybe
+  , sqrtModSharp
   , sqrtModF
   , sqrtModFList
   ) where
 
+import Safe (headMay)
+
 import Control.Monad (liftM2)
 import Data.Bits
 import Data.List (nub)
 import GHC.Integer.GMP.Internals
 
+import Math.NumberTheory.GCD (coprime)
 import Math.NumberTheory.Moduli.Chinese
 import Math.NumberTheory.Moduli.Jacobi
 import Math.NumberTheory.Primes.Sieve (sieveFrom)
+import Math.NumberTheory.Primes.Testing.Probabilistic (isPrime)
+import Math.NumberTheory.Primes.Factorisation.Montgomery (factorise)
 import Math.NumberTheory.Utils (shiftToOddCount, splitOff)
 
+-- | Extended flag structure to check if square root is computed in finite field.
+data FieldCharacterictic = Invalid
+                         | Prime
+                         | PrimePower Integer Int
+  deriving Show
+
+-- | Check if value is power of some prime.
+checkPrimePower :: Integer -> Maybe (Integer, Int)
+checkPrimePower = headMay . factorise
```

Comment:
It should be faster to use `Math.NumberTheory.Powers.General.highestPower` here.